feat: allow user to pass multiple files to `fix` and `check` commands

sinrohit-desco opened the pull request:
Resolves https://github.com/oppiliappan/statix/issues/118

mightyiam commented:
I think we should have this. How about a test though?

TLATER reviewed (commented):
Yeah, I much prefer this, I overlooked the `Walker` when I looked at this.

I'll close #96

neko-desnix commented:
I completely implemented the iterator using Walk from ignore crate. Let me know your thoughts @TLATER @mightyiam
If everything looks good to you guys, I will add tests.

TLATER reviewed (commented):
Looks *really* good, much better than the existing code! Error handling might need some work though.

To be clear, since apparently this is less common parlance than I thought, "nit" means a small thing that I don't *really* think needs to be addressed but point out anyway because I'm overly specific about programming. So feel free to ignore those comments if what they're talking about doesn't also annoy you.

TLATER commented on bin/src/dirs.rs:
```diff
@@ -1,113 +1,63 @@
-    Ok(walker.filter(|path: &PathBuf| matches!(path.extension(), Some(e) if e == "nix")))
+        for pattern in ignore_patterns {
+            override_builder
+                .add(&format!("!{}", pattern))
```
If changing the behavior is acceptable, I'd suggest removing the current flag and adding the `--include` and `--exclude` flags that the `ignore` crate suggests.

neko-desnix replied:
I didn't change the behavior, I just reimplemented things to make it simple. But this is what I came up with, I could go over it again and do some more refactoring.

TLATER reviewed (approved):
LGTM, no further comments :) I'd love to resolve my earlier ones, sadly I don't have permissions to do that.

mightyiam commented:
Sorry, I think #138 is the priority so we've been working on that.

mightyiam commented:
Hey, #138 is complete. Sorry for the conflicts. Looking forward to reviewing this again.

sinrohit-desco commented:
> Hey, #138 is complete. Sorry for the conflicts. Looking forward to reviewing this again.

Hi Shahar! I have rebased the PR and resolved the conflicts.

mightyiam commented on bin/src/config.rs:
```diff
@@ -5,7 +5,7 @@ use std::{
 };
 
-use crate::{LintMap, dirs, err::ConfigErr, utils};
+use crate::{dirs, err::ConfigErr, utils, LintMap};
```
Please do revert.

mightyiam reviewed (commented):
Should this new feature include a test?

voruti commented:
I was just testing this out, but it's not compiling correctly.

```log
error: this argument is passed by value, but not consumed in the function body
  --> bin/src/dirs.rs:15:22
error: variables can be used directly in the `format!` string
  --> bin/src/dirs.rs:43:23
  --> bin/src/dirs.rs:60:17
```

sinrohit-desco commented:
Apologies for leaving it unattended for this long. 

> I was just testing this out, but it's not compiling correctly.
> 
> ```
> error: this argument is passed by value, but not consumed in the function body
>   --> bin/src/dirs.rs:15:22
> error: variables can be used directly in the `format!` string
>   --> bin/src/dirs.rs:43:23
>   --> bin/src/dirs.rs:60:17
> ```

That's the clippy warning (error). I ran cargo build and didn't run clippy checks.  Let me fix that and also rebase.

sinrohit-desco committed:
feat: allow user to pass multiple files to fix and check commands

reimplement the iterator using ignore::Walk

This implementation simplifies things a bit also it returns a lazy
iterator which is memory efficient for large codebases.

mightyiam commented:
Thank you for resuming. But, would you mind submitting against https://github.com/molybdenumsoftware/statix ?

sinrohit-desco commented:
> Thank you for resuming. But, would you mind submitting against https://github.com/molybdenumsoftware/statix ?

Sure thing!!

sinrohit-desco commented:
Closing in favour of https://github.com/molybdenumsoftware/statix/pull/1921

> Thank you for resuming. But, would you mind submitting against https://github.com/molybdenumsoftware/statix ?
